### DISCUSSION TITLE
Should we allow parties to enter Agreements with different lock amounts?

### DISCUSSION THREAD

**Discussion by luisivan** [nation3/court#6] [Ideas]
As pointed out by @gregfromstl:

> Another thing to note: we should allow the default unlock of each person to swap to allow for payment agreements. For example I want to pay someone 5 NATION for a job, and we agree to lock 2 NATION for the agreement. I lock 7 NATION and they lock 2 NATION, then when the agreement completes, they get back 7 NATION and I get back 2 (the 5 NATION changes hands)

**Comment by luisivan** [comment]
A way to incorporate this without created added complexity is for parties to be able to enter an Agreement with `>= lockAmount`. In the example mentioned in the comment above, the party carrying out the job would only enter the Agreement once the other party has locked the base amount (2) plus the payment amount (5).

That way we don't need to incorporate a system for an Agreement to specify specific locks per address, which can get complex. Thoughts?

**Reply by xPi2** [reply]
For the base agreement primitive the method to enter should allow you to lock whatever amount >= required lock amount. Specific implementations could extend this mechanism for their specific needs.

**Reply by gregfromstl** [reply]
How do we handle the distribution then? In the example above, we could allow locking more than the required amount, but we also have to indicate that the excess amount should be sent to the other user upon contract resolution. I do think this should be a unique implementation, and likely not something we'd need in the primitive.

**Reply by xPi2** [reply]
In the primitive agreement we should allow to lock whatever amount and store that value, each specific agreement could handle the termination / punishment procedure to handle the funds on its own.

I think we should define like 3 different initial use cases that we can implement as specific agreements so we can see how much can we abstract the agreement primitive.
